---
title: "Dio3 Hypomorph"
author: "Clayton Pio Santiago"
date: "2023-06-05"
output: html_document
---
Load required packages
```{r include=FALSE}
library(Seurat)
library(Signac)
library(tidyverse)
library(viridis)
library(RColorBrewer)
library(monocle3)
library(harmony)
library(SeuratWrappers)
library(patchwork)
library(plotly)
library(tidyseurat)
library(scDblFinder)
library(tradeSeq)
library(ArchR)
library(slingshot)
library(pheatmap)
library(ComplexHeatmap)
library(cowplot)
library(scales)
library(UCell)
library(Matrix)
library(SingleCellExperiment)
library(apeglm)
library(BiocParallel)
library(nichenetr)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TFBSTools)
library(JASPAR2020)
set.seed(1234)
```

Load RNA counts from 10X multiome data 
```{r include=FALSE}
scM144R.data <- Read10X(data.dir = "D:/R_files/Christina_boblab/Cellranger/scM144R/filtered_feature_bc_matrix/")
scM145R.data <- Read10X(data.dir = "D:/R_files/Christina_boblab/Cellranger/scM145R/filtered_feature_bc_matrix/")
```

Create Seurat object
```{r include=FALSE}
scM144R <- CreateSeuratObject(counts = scM144R.data, project = "con_1", min.cells = 3, min.features = 200)
scM145R <- CreateSeuratObject(counts = scM145R.data, project = "dio3_hypo_1", min.cells = 3, min.features = 200)
rm(list=ls(pattern=".data"))
```

Add metadata
```{r include=FALSE}
scM144R@meta.data$age <- "200day"
scM145R@meta.data$age <- "200day"

scM144R@meta.data$genotype <- "Control"
scM145R@meta.data$genotype <- "DIO3_hypo"

scM144R@meta.data$run <- "Nuclei"
scM145R@meta.data$run <- "Nuclei"

scM144R@meta.data$seqID <- "scM144"
scM145R@meta.data$seqID <- "scM145"
```

Merge cell and nuclei datasets respectively and remove individual files
```{r}
dio3 <- merge(scM144R, y = c(scM145R), 
              add.cell.ids = c("scM144","scM145"))
rm(list=ls(pattern="scM"))
```

Order meta data
```{r}
order1<- c("Control","DIO3_hypo")
dio3@meta.data$genotype <- factor(dio3@meta.data$genotype, levels = order1)
order2<- c("con_1","dio3_hypo_1")
dio3@meta.data$orig.ident <- factor(dio3@meta.data$orig.ident, levels = order2)
rm(list=ls(pattern="ord"))
```

Add percent mitocondrial and seurat cell cycle scores
```{r include=FALSE}
dio3[["percent.mt"]] <- PercentageFeatureSet(dio3, pattern = "^MT-")
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
dio3 <- CellCycleScoring(dio3, s.features = s.genes, g2m.features = g2m.genes, set.ident = F)
rm (g2m.genes, s.genes)
```

Add number of genes per UMI for each cell to metadata
```{r}
dio3@meta.data$log10GenesPerUMI <- log10(dio3@meta.data$nFeature_RNA)/log10(dio3@meta.data$nCount_RNA)
```

Scatter plot to show QC metrics
```{r, echo=FALSE, message=FALSE, results='hide'}
FeatureScatter(dio3, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident")
FeatureScatter(dio3, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident")
```

Violin plot to show QC metrics
```{r, echo=FALSE, message=FALSE, results='hide'}
VlnPlot(dio3, features = c("nFeature_RNA", "nCount_RNA","percent.mt"), group.by = "orig.ident", ncol = 3)
```

Boxplot to show QC
```{r}
dio3@meta.data %>%
  ggplot(aes(x=orig.ident, y=(nFeature_RNA), fill=orig.ident)) + 
  geom_boxplot() + theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold")) + ggtitle("Single Nuclei Gene Boxplot")
```

Correlation between genes and UMI
```{r message=FALSE, warning=FALSE}
dio3@meta.data %>%
  ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  geom_point() + scale_colour_gradient(low = "gray90", high = "red") +
  stat_smooth(method=lm) + scale_x_log10() + scale_y_log10() + 
  theme_classic() + facet_wrap(~orig.ident)
```

Visualizing the cell densities of genes detected per UMI, percent.mt, UMIs and genes
```{r}
dio3@meta.data %>% 
  ggplot(aes(color=orig.ident, x=log10GenesPerUMI, fill= orig.ident)) + 
  geom_density(alpha = 0.2) + theme_classic() +  ylab("Cell density") +  geom_vline(xintercept = 0.75)
dio3@meta.data %>% 
  ggplot(aes(color=orig.ident, x=percent.mt, fill= orig.ident)) + 
  geom_density(alpha = 0.2) + theme_classic() +  ylab("Cell density") +  geom_vline(xintercept = 3) + coord_cartesian(xlim = c(0,20),ylim = c(0,5))
dio3@meta.data %>% 
  ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  geom_density(alpha = 0.2) + theme_classic() +  ylab("Cell density") +  geom_vline(xintercept = 40000)
dio3@meta.data %>% 
  ggplot(aes(color=orig.ident, x=nFeature_RNA, fill= orig.ident)) + 
  geom_density(alpha = 0.2) + theme_classic() +  ylab("Cell density") 
```
Number of cells before filtering
```{r}
table (dio3@meta.data$orig.ident)
```

Save files and Filter out low quality cells
```{r}
saveRDS(dio3, file = "F:/DIO3_organoid/R_files/dio3_nofiltering.rds")
dio3 <- subset(x = dio3, subset= percent.mt < 3)
dio3 <- subset(x = dio3, subset= nCount_RNA > 900)
saveRDS(dio3, file = "F:/DIO3_organoid/R_files/dio3_highmtrem.rds")
```

Number of cells after filtering
```{r}
table (dio3@meta.data$orig.ident)
```

Identify doublets
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
sce <- as.SingleCellExperiment(dio3)
sce <- scDblFinder(sce, samples="seqID",BPPARAM=SerialParam(RNGseed = 1234)) #compatible with windows machines
dio3@meta.data$doublet <- plyr::mapvalues(x = rownames(dio3@meta.data), 
                                           from = rownames(colData(sce)), to = as.character(sce$scDblFinder.class))
dio3@meta.data$doublet <- factor(dio3@meta.data$doublet, levels = c("singlet","doublet"))
dio3@meta.data$doublet_score <- plyr::mapvalues(x = rownames(dio3@meta.data), 
                                                 from = rownames(colData(sce)), to = sce$scDblFinder.score)
dio3@meta.data$doublet_score<-as.numeric(dio3@meta.data$doublet_score)
rm (sce)
```

Number of called singlets and doublets and genes in the dataset
```{r message=FALSE, warning=FALSE}
cbind(table(dio3$orig.ident, dio3$doublet), table(dio3$orig.ident, dio3$nCount_RNA > 40000))
```

Filter out doublets and cells with high UMI
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
dio3 <- subset(x = dio3, subset= nCount_RNA < 40000)
dio3 <- subset(x = dio3, subset= doublet == "singlet")
saveRDS(dio3, file = "F:/DIO3_organoid/R_files/dio3_filtered.rds")
```

Number of cells after filtering
```{r}
table (dio3@meta.data$orig.ident)
```

Bar graph to show UMI and genes for each cell
```{r}
comb <- dio3@meta.data
comb$rows <- rownames(comb)
comb <-comb[order(comb$orig.ident),]

ggbarplot(comb, x = "rows", y = "nFeature_RNA", color = "orig.ident", fill = "orig.ident",sort.val = "desc",sort.by.groups = TRUE) +  scale_y_continuous(expand = c(0, 0), trans = "log10") + theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
ggbarplot(comb, x = "rows", y = "nCount_RNA", color = "orig.ident", fill = "orig.ident",sort.val = "desc",sort.by.groups = TRUE) + scale_y_continuous(expand = c(0, 0), trans = "log10") + theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
rm (comb)
```

Mean UMI and genes in the dataset
```{r}
dio3@meta.data %>%  group_by(orig.ident) %>%
    summarise(mean_nCount_RNA = mean(nCount_RNA),
            mean_nFeature_RNA = mean(nFeature_RNA),
            median_nCount_RNA = median(nCount_RNA),
            median_nFeature_RNA = median(nFeature_RNA))
```

Integrate the data
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
dio3 <- NormalizeData(dio3) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA(verbose = FALSE)
dio3 <- RunHarmony(dio3, group.by.vars = "orig.ident")
dio3 <- RunUMAP(dio3, reduction = "harmony", dims = 1:20)
dio3 <- RunUMAP(dio3, reduction = "harmony", dims = 1:20, 
                reduction.name = "umap3d", n.components = 3) #3D UMAP
dio3 <- FindNeighbors(dio3, reduction = "harmony", dims = 1:20)
dio3 <- FindClusters(dio3, resolution = 1.2)
DefaultAssay(dio3) <- "RNA"
saveRDS(dio3, file = "F:/DIO3_organoid/R_files/dio3_filtered.rds")
```

Name Clusters and remove additional doublets
```{r}
seu_clust<-c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18",
            "19","20","21","22","23","24")
id <-c("Cones","Muller glia ","Cones","Rods","nRPCs","Rods","Cones","RPCs","Glia-like",
"Bipolar","RPE","Glia-like","Amacrine","PR Precursors","Doublets","PR Precursors","Astrocytes",
"RPCs","RPCs","Amacrine","BSLC","Horizontal","RPE","Cycling RPCs","Choroid plexus")
dio3@meta.data$Celltype <- plyr::mapvalues(x = dio3@meta.data$seurat_clusters, from = seu_clust, to = id)

order1<-c("Rods","Cones","Bipolar","Amacrine","Horizontal","PR Precursors","nRPCs",
"Muller glia ","RPE","RPCs","Cycling RPCs","Glia-like","Astrocytes","BSLC","Choroid plexus",
"Doublets")
dio3@meta.data$Celltype <- factor(dio3@meta.data$Celltype, levels = order1)
Idents(dio3) <- "Celltype"

dio3 <- subset (dio3, idents = "Doublets", invert = T)
rm (seu_clust, id, order1)
```

Number of cells after filtering
```{r}
table (dio3@meta.data$orig.ident)
```

Plot number OR percent of cells per cluster for each sample
```{r message=FALSE, warning=FALSE,  results='hide'}
meta.data <- dio3@meta.data
counts <- group_by(meta.data, genotype, Celltype)	%>% dplyr::summarise(count = n()) %>%
    group_by(genotype) %>% dplyr::mutate(clust_total = sum(count)) %>% dplyr::mutate(clust_prop = count / clust_total * 100)
ggplot(counts, aes(Celltype, clust_prop, fill = genotype)) + 
    geom_col(position = position_dodge2(preserve = "single")) + 
    labs(y = "Cell Percentage") + theme_classic() + RotatedAxis()
rm (meta.data,counts)
```

Density Plot to show celltype differences between genotypes
```{r}
mincell <- min(table (dio3@meta.data$genotype))
Idents (dio3) <- "genotype"
down <- subset(dio3, downsample = mincell)
Idents (down) <- "Celltype"
down <- subset(down, downsample = 500) #Control for celltype
tidyseurat::ggplot(down, aes(x= umap_1, y= umap_2) ) +
        stat_density2d(aes(fill= ..density..), geom= "raster", contour = F) +
        scale_x_continuous(expand = c(0, 0)) +
        scale_y_continuous(expand = c(0, 0)) +
        scale_fill_viridis_c(option="magma") +
        labs(fill = "Density") +
        facet_wrap("genotype") +
        theme(text= element_text(size= 14) ,
              legend.position= 'right',
              strip.background = element_rect(color="white", fill="white"))
rm (down, mincell)
```

Subset data
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
pr <- subset (dio3, idents = c("Rods","Cones","PR Precursors","nRPCs"))
rem <- CellSelector(DimPlot(pr)) #Remove 3 cells
Idents(pr, cells = rem) <- "rem"
pr <- subset (pr, cells = rem, invert = T)
rm (rem)

sce <- as.SingleCellExperiment(pr, assay = "RNA")
sce <- slingshot(sce, reducedDim = 'UMAP', clusterLabels = colData(sce)$ident, start.clus = "nRPCs")

cols <- c("#FC717F","#39B600","#00A5FF","#C77CFF")
plotcol <- cols[(sce$ident)]
plot(reducedDims(sce)$UMAP, col = plotcol, pch=16, asp = 0, cex = 0.5)
lines(SlingshotDataSet(sce), linInd = 1, lwd=3, col='red')
lines(SlingshotDataSet(sce), linInd = 2, lwd=3, lty=2, col='black')
rm(cols, plotcol)

pseudotime <- slingPseudotime(SlingshotDataSet(sce), na = TRUE)
pr@meta.data$cone_pseudotime <- pseudotime[,1]
pr@meta.data$rod_pseudotime <- pseudotime[,2]
rm(pseudotime)

df <- as.data.frame(pr@meta.data) %>% dplyr::filter(!is.na(cone_pseudotime))
ggplot(df, aes(x=cone_pseudotime, color = genotype, fill= genotype)) + 
  geom_density(alpha = 0.2) + 
  theme_classic() +  
  scale_color_manual(values=c("#F8766D","#00BFC4")) +
  scale_fill_manual(values=c("#F8766D","#00BFC4"))

df <- as.data.frame(pr@meta.data) %>% dplyr::filter(!is.na(rod_pseudotime))
ggplot(df, aes(x=rod_pseudotime, color = genotype, fill= genotype)) + 
  geom_density(alpha = 0.2) + 
  theme_classic() +  
  scale_color_manual(values=c("#F8766D","#00BFC4")) +
  scale_fill_manual(values=c("#F8766D","#00BFC4"))
rm (df)
```

S and M cone scoring
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "RNA"
scone<-c("OPN1SW","CCDC136","PCP4","GNGT1","SAG","TTR","CADM3","NDNF","PRKAR2B","TBX2","YBX3")
mcone<-c("PDE6H","OPN1MW3","GUCA1A","VOPP1","SLC25A25","GRK7","PPA1","CA2","SAMD7")
rod <- c("SAG","PDE6A","NR2E3","GNAT1","ROM1","NRL","GNGT1","PDE6B","REEP6","RHO","CABP5")
pr <- AddModuleScore_UCell(pr, features = list(scone), name = "scone_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1scone_score'] <- 'scone_score'
pr <- AddModuleScore_UCell(pr, features = list(mcone), name = "mcone_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1mcone_score'] <- 'mcone_score'
pr <- AddModuleScore_UCell(pr, features = list(rod), name = "rod_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1rod_score'] <- 'rod_score'
rm (scone,mcone,rod)
```

Scoring for various cell death pathways and TH signaling
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "RNA"
# KEGG_REGULATION_OF_AUTOPHAGY (hsa04140)
autophagy<- c("ATG7","GABARAP","GABARAPL2","ATG4A","ATG4B","GABARAPL1",
"ULK3","PIK3R4","IFNA1","IFNA2","IFNA4","IFNA5","IFNA6","IFNA7","IFNA8",
"IFNA10","IFNA13","IFNA14","IFNA16","IFNA17","IFNA21","IFNG","INS","BECN2",
"PIK3C3","PRKAA1","PRKAA2","ATG3","ULK1","ATG4C","ATG4D","BECN1","ATG12","ATG5",
"ULK2")

# BIOCARTA_DEATH_PATHWAY 
death_pathway<- c("BCL2","TNFRSF25","CASP8","CFLAR","SPTAN1","GAS2","RELA",
"APAF1","BIRC3","NFKB1","BIRC2","XIAP","TNFSF10","BID","CASP10","CASP6",
"CASP7","CASP9","LMNA","CHUK","TRADD","RIPK1","FADD","MAP3K14","CASP3","DFFA",
"DFFB","NFKBIA","TRAF2")

# KEGG_APOPTOSIS (hsa04210)
apotosis<- c("AKT3","IRAK3","CHP1","CHUK","CSF2RB","DFFA","DFFB","ENDOG",
"AKT1","AKT2","ENDOD1","PIK3R5","APAF1","BIRC2","BIRC3","XIAP","FAS","IKBKB",
"IL1A","IL1B","IL1R1","IL1RAP","FASLG","IL3","IL3RA","IRAK1","IRAK2","MYD88",
"ATM","NFKB1","NFKBIA","NGF","NTRK1","IRAK4","PIK3CA","PIK3CB","PIK3CD","PIK3CG",
"PIK3R1","PIK3R2","CYCS","PPP3CA","PPP3CB","PPP3CC","PPP3R1","PPP3R2","PRKACA",
"PRKACB","PRKACG","PRKAR1A","PRKAR1B","PRKAR2A","PRKAR2B","PRKX","BAD","BAX",
"BCL2","RELA","BCL2L1","BID","CHP2","TNF","TNFRSF1A","TP53","TRAF2","CAPN1",
"CAPN2","CASP3","CASP6","CASP7","CASP8","CASP9","CASP10","PIK3R3","IKBKG",
"TRADD","RIPK1","TNFSF10","FADD","TNFRSF10D","TNFRSF10C","TNFRSF10B","TNFRSF10A",
"CFLAR","MAP3K14","AIFM1","EXOG")

# Response to Thyroid Hormone (GO:0097066)
th_signal<- c("CAB39","SLC34A1","GBA","STK11","TOMM70","GBA1","HPN","C2",
"CTSL","CTSB","F7","INHBB","CTSH","KIT","THRB","ACE","GATA1","CTSS","LMO2",
"RDX","GCLC","GCLM","SLC26A5","KLF9","HES1","MED1","GHSR","GAS2L1","BRD8","CRLS1")

pr <- AddModuleScore_UCell(pr, features = list(autophagy), name = "autophagy_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1autophagy_score'] <- 'autophagy_score'
pr <- AddModuleScore_UCell(pr, features = list(death_pathway), name = "death_pathway_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1death_pathway_score'] <- 'death_pathway_score'
pr <- AddModuleScore_UCell(pr, features = list(apotosis), name = "apotosis_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1apotosis_score'] <- 'apotosis_score'
pr <- AddModuleScore_UCell(pr, features = list(th_signal), name = "th_signal_score", assay = "RNA")
names(pr@meta.data)[names(pr@meta.data) == 'signature_1th_signal_score'] <- 'th_signal_score'
rm (autophagy,death_pathway,apotosis,th_signal)
VlnPlot(pr, features = c("autophagy_score","death_pathway_score","apotosis_score"))
```

Rename PR clusters
```{r}
seu_clust<-c("0","2","3","4","5","6","13","15")
id <-c("ML_cone","S_cone","Rods","nRPCs","Rods","ML_cone","PR Precursors","PR Precursors")
pr@meta.data$Celltype <- plyr::mapvalues(x = pr@meta.data$seurat_clusters, from = seu_clust, to = id)

pr@meta.data$genotype_celltype <- paste(pr@meta.data$genotype,
                                        pr@meta.data$Celltype, sep = "_")
order1 <- c("Control_Rods","DIO3_hypo_Rods","Control_S_cone",
            "DIO3_hypo_S_cone","Control_ML_cone","DIO3_hypo_ML_cone",
            "Control_PR Precursors","DIO3_hypo_PR Precursors",
            "Control_nRPCs","DIO3_hypo_nRPCs")
pr@meta.data$genotype_celltype <- factor(pr@meta.data$genotype_celltype, levels = order1)

order2<-c("Rods","S_cone","ML_cone","PR Precursors","nRPCs")
pr@meta.data$Celltype <- factor(pr@meta.data$Celltype, levels = order2)
Idents(pr) <- "Celltype"
rm (seu_clust, id, order1, order2)

dio3@meta.data <- dio3@meta.data %>%
  mutate(Celltype2 = coalesce(pr@meta.data$Celltype[match(rownames(dio3@meta.data), 
                                                          rownames(pr@meta.data))], dio3@meta.data$Celltype))
dio3@meta.data$genotype_celltype2 <- paste(dio3@meta.data$genotype,
                                        dio3@meta.data$Celltype2, sep = "_")
```

For 3D plotting
```{r}
genes <- c("")
fetch<- FetchData(object = dio3, 
                 vars = c("umap_1", "umap_2", "umap_3", 
                          "genotype", "seurat_clusters", "Celltype"), layer = "data")

# Plot gene expression on 3D UMAP
plot_ly(data = fetch, x = ~umap_1, y = ~umap_2, z = ~umap_3, 
        type = "scatter3d", mode = "markers",
        marker = list(size = 1, width=1,  
                      color =fetch$Ascl1, showscale = T)) 
# Plot cell identity on 3D UMAP
plotly::plot_ly(data = fetch, x = ~"umap_1", y = ~"umap_2", z = ~"umap_3", 
        type = "scatter3d", mode = "markers", 
        color = ~"genotype",
        marker = list(size = 1, width=1))
```

ATAC data
Load ATAC counts from 10X multiome data 
######REWORK#####
IMPROVE TO ADD COUNTS DIRECTLY TO ORIGINAL OBJECT
```{r include=FALSE}
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevels(annotation) <- paste0('chr', seqlevels(annotation))

scM144.data <- Read10X(data.dir = "H:/Bob_lab/Christina_dio3/Raw/Control/raw_feature_bc_matrix/")
scM145.data <- Read10X(data.dir = "H:/Bob_lab/Christina_dio3/Raw/Dio3/raw_feature_bc_matrix/")
scM144 <- CreateSeuratObject(counts = scM144.data$`Gene Expression`, project = "con_1")
scM145 <- CreateSeuratObject(counts = scM145.data$`Gene Expression`, project = "dio3_hypo_1")
scM144[["Peaks"]] <- CreateChromatinAssay(counts=scM144.data$Peaks,sep = c(":", "-"),
        fragments = 'H:/Bob_lab/Christina_dio3/Raw/Control/atac_fragments.tsv.gz',
        annotation = annotation)
scM145[["Peaks"]] <- CreateChromatinAssay(counts=scM145.data$Peaks,sep = c(":", "-"),
        fragments = 'H:/Bob_lab/Christina_dio3/Raw/Dio3/atac_fragments.tsv.gz',
        annotation = annotation)
rm(list=ls(pattern=".data"))

d3 <- merge(scM144, y = c(scM145), add.cell.ids = c("scM144","scM145"))
d3 <- subset(d3, cells = rownames(dio3@meta.data))
dio3[["Peaks10x"]] <- d3[["Peaks"]]
rm (d3, scM144, scM145)
```

Call celltype specific peaks using MACS2 and create new Peak assay
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(dio3) <- "Peaks10x"
#Need to run CallPeaks in a linux environment
#Change fragment path if moving Signac object
peaks <- CallPeaks(dio3, group.by = "Celltype", macs2.path = "/zp1/data/Blackshaw/mods/envs/PeakCalling_analysis/bin/macs2")
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

counts <- FeatureMatrix(fragments = Fragments(dio3),features = peaks, cells = colnames(dio3))
dio3[["Peaks"]] <- CreateChromatinAssay(counts = counts, fragments = frags,  annotation = annotation)
DefaultAssay(dio3) <- "Peaks10x"

saveRDS(dio3, file = "F:/Bob_lab/Christina_dio3/R_files/dio3_filtered_macs2_int.rds")

rm (annotation, peaks, counts)
```

Run standard ATAC pipeline
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(dio3) <- "Peaks"
dio3 <- NucleosomeSignal(dio3) %>% TSSEnrichment()
dio3 <- FindTopFeatures(dio3) %>% RunTFIDF %>% RunSVD
```

Compute QC metrics
######REWORK#####
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
scM144 <- CountFragments("H:/Bob_lab/Christina_dio3/Raw/Control/atac_fragments.tsv.gz")
scM145 <- CountFragments("H:/Bob_lab/Christina_dio3/Raw/Dio3/atac_fragments.tsv.gz")
scM144$CB <- paste0("scM144_", scM144$CB)
scM145$CB <- paste0("scM145_", scM145$CB)
combined_df <- rbind(scM144, scM145)
combined_df <- combined_df[combined_df$CB %in% rownames(dio3@meta.data), ]
rownames(combined_df) <- combined_df$CB
rownames(combined_df) <- fragment$CB
dio3@meta.data$fragments <- combined_df[colnames(dio3), "reads_count"]
dio3 <- FRiP(dio3, assay = 'Peaks', total.fragments = "fragments")
rm (scM144, scM145, combined_df)
```

Plot QC metrics
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DensityScatter(dio3, x = 'nCount_Peaks', y = 'TSS.enrichment', log_x = TRUE, quantiles = TRUE)
FragmentHistogram(dio3, group.by = 'genotype', region = 'chr1-1-10000000')
TSSPlot(dio3, group.by = 'genotype') + NoLegend()
VlnPlot(dio3,  features = c('FRiP', "nCount_Peaks",'TSS.enrichment', 
                    'nucleosome_signal'),  pt.size = 0.1, ncol = 4, group.by = 'genotype')
```

Filter out cells based on ATAC QC metrics
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
dio3 <- subset(dio3, subset = nCount_Peaks < 50000 & 
                  nCount_Peaks > 1000 &  nucleosome_signal < 2 &  TSS.enrichment > 1)
```

Number of cells after filtering
```{r}
table (dio3@meta.data$orig.ident)
```

Integrate data based on RNA and ATAC
#Not required
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(dio3) <- "Peaks"
dio3 <- RunHarmony(dio3, group.by.vars = "orig.ident", reduction = "lsi", 
                   assay.use = "Peaks", reduction.save = "harmony.lsi", project.dim = F)
dio3 <- RunUMAP(dio3, reduction = 'harmony.lsi', dims = 2:30, reduction.name = "umap_atac")
dio3 <- FindMultiModalNeighbors(dio3, reduction.list = list("harmony", "harmony.lsi"), 
                                dims.list = list(1:20, 2:30), 
                                modality.weight.name = "RNA.weight",  verbose = T)
dio3 <- RunUMAP(dio3,nn.name = "weighted.nn",assay = "RNA",
                reduction.name = "umap_wnn")
```

Create a gene activity matrix and link ATAC peaks to gene expression
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
gene.activities <- GeneActivity(dio3)
dio3[['ATAC']] <- CreateAssayObject(gene.activities)
dio3 <- NormalizeData(dio3, assay = 'ATAC',
  normalization.method = 'LogNormalize',scale.factor = median(dio3$nCount_ATAC))

dio3 <- RegionStats(dio3, genome = BSgenome.Hsapiens.UCSC.hg38)
dio3 <- LinkPeaks(dio3, peak.assay = "Peaks", expression.assay = "RNA") #subsetting will remove links
rm (gene.activities)
```

Add motif information and compute motif activity
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
pwm <- getMatrixSet(JASPAR2020, opts = list(collection = "CORE", 
                        tax_group = 'vertebrates', all_versions = FALSE))
dio3 <- AddMotifs(dio3, genome = BSgenome.Hsapiens.UCSC.hg38, pfm = pwm)
dio3 <- RunChromVAR(dio3, genome = BSgenome.Hsapiens.UCSC.hg38)
rm (pwm)
```

Find DA peaks and motifs in control S cones
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
motif_list <- as.data.frame (unlist(dio3@assays$Peaks@motifs@motif.names))
motif_list$id <- rownames (motif_list)
colnames(motif_list)[1] <- "name"

DefaultAssay(pr) <- "Peaks"
con_scone_peak <- FindMarkers(pr, ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "S_cone", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.conscone.peak <- rownames(con_scone_peak[con_scone_peak$p_val  < 0.005, ])
annot = ClosestFeature(pr, region = rownames(con_scone_peak))
con_scone_peak <- con_scone_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("S_cone"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.conscone.peak, ],  n = 50000)

conscone.enriched.motifs <- FindMotifs(pr, features = top.conscone.peak, 
                              background = peaks.matched)
conscone.enriched.motifs <- subset(conscone.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
conscone.diff_act <- FindMarkers(pr,ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "S_cone", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
conscone.diff_act$motif_name <- plyr::mapvalues(x =rownames(conscone.diff_act), 
                                       from = motif_list$id, to = motif_list$name)
write.csv(con_scone_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/con_scone_peak.csv")
write.csv(conscone.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/conscone.enriched.motifs.csv")
write.csv(conscone.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/conscone.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.conscone.peak,annot)
```

Find DA peaks and motifs in control ML cones
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "Peaks"
con_MLcone_peak <- FindMarkers(pr, ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "ML_cone", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.conMLcone.peak <- rownames(con_MLcone_peak[con_MLcone_peak$p_val  < 0.005, ])
annot = ClosestFeature(pr, region = rownames(con_MLcone_peak))
con_MLcone_peak <- con_MLcone_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("ML_cone"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.conMLcone.peak, ],  n = 50000)

conMLcone.enriched.motifs <- FindMotifs(pr, features = top.conMLcone.peak, 
                              background = peaks.matched)
conMLcone.enriched.motifs <- subset(conMLcone.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
conMLcone.diff_act <- FindMarkers(pr,ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "ML_cone", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
conMLcone.diff_act$motif_name <- plyr::mapvalues(x =rownames(conMLcone.diff_act), 
                                       from = motif_list$id, to = motif_list$name)

write.csv(con_MLcone_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/con_MLcone_peak.csv")
write.csv(conMLcone.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/conMLcone.enriched.motifs.csv")
write.csv(conMLcone.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/conMLcone.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.conMLcone.peak,annot)
```

Find DA peaks and motifs in control Rods
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "Peaks"
con_rod_peak <- FindMarkers(pr, ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "Rods", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.conrod.peak <- rownames(con_rod_peak[con_rod_peak$p_val  < 0.005, ])
annot = ClosestFeature(pr, region = rownames(con_rod_peak))
con_rod_peak <- con_rod_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("Rods"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.conrod.peak, ],  n = 50000)

conrod.enriched.motifs <- FindMotifs(pr, features = top.conrod.peak, 
                              background = peaks.matched)
conrod.enriched.motifs <- subset(conrod.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
conrod.diff_act <- FindMarkers(pr,ident.1 = "Control", ident.2 = "DIO3_hypo", 
                         group.by = "genotype", subset.ident = "Rods", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
conrod.diff_act$motif_name <- plyr::mapvalues(x =rownames(conrod.diff_act), 
                                       from = motif_list$id, to = motif_list$name)

write.csv(con_rod_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/con_rod_peak.csv")
write.csv(conrod.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/conrod.enriched.motifs.csv")
write.csv(conrod.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/conrod.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.conrod.peak,annot)
```

Find DA peaks and motifs in DIO3 hypomorph S cones
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "Peaks"
dio3_scone_peak <- FindMarkers(pr, ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "S_cone", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.dio3scone.peak <- rownames(dio3_scone_peak[dio3_scone_peak$p_val < 0.005, ])
annot = ClosestFeature(pr, region = rownames(dio3_scone_peak))
dio3_scone_peak <- dio3_scone_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("S_cone"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.dio3scone.peak, ],  n = 50000)

dio3scone.enriched.motifs <- FindMotifs(pr, features = top.dio3scone.peak, 
                              background = peaks.matched)
dio3scone.enriched.motifs <- subset(dio3scone.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
dio3scone.diff_act <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "S_cone", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
dio3scone.diff_act$motif_name <- plyr::mapvalues(x =rownames(dio3scone.diff_act), 
                                       from = motif_list$id, to = motif_list$name)

write.csv(dio3_scone_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3_scone_peak.csv")
write.csv(dio3scone.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3scone.enriched.motifs.csv")
write.csv(dio3scone.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3scone.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.dio3scone.peak,annot)
```

Find DA peaks and motifs in DIO3 hypomorph ML cones
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "Peaks"
dio3_MLcone_peak <- FindMarkers(pr, ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "ML_cone", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.dio3MLcone.peak <- rownames(dio3_MLcone_peak[dio3_MLcone_peak$p_val < 0.005, ])
annot = ClosestFeature(pr, region = rownames(dio3_MLcone_peak))
dio3_MLcone_peak <- dio3_MLcone_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("ML_cone"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.dio3MLcone.peak, ],  n = 50000)

dio3MLcone.enriched.motifs <- FindMotifs(pr, features = top.dio3MLcone.peak, 
                              background = peaks.matched)
dio3MLcone.enriched.motifs <- subset(dio3MLcone.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
dio3MLcone.diff_act <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "ML_cone", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
dio3MLcone.diff_act$motif_name <- plyr::mapvalues(x =rownames(dio3MLcone.diff_act), 
                                       from = motif_list$id, to = motif_list$name)

write.csv(dio3_MLcone_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3_MLcone_peak.csv")
write.csv(dio3MLcone.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3MLcone.enriched.motifs.csv")
write.csv(dio3MLcone.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3MLcone.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.dio3MLcone.peak,annot)
```

Find DA peaks and motifs in DIO3 hypomorph rods
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
DefaultAssay(pr) <- "Peaks"
dio3_rod_peak <- FindMarkers(pr, ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "Rods", 
                         assay = "Peaks", test.use = "LR", 
                         min.pct = 0.05, latent.vars = "nCount_Peaks", only.pos = T)
top.dio3rod.peak <- rownames(dio3_rod_peak[dio3_rod_peak$p_val < 0.005, ])
annot = ClosestFeature(pr, region = rownames(dio3_rod_peak))
dio3_rod_peak <- dio3_rod_peak %>% mutate(closest_gene = annot$gene_name,
                                            distance_gene = annot$distance)

open.peaks <- AccessiblePeaks(pr, idents = c("Rods"))
meta.feature <- GetAssayData(pr, assay = "Peaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ],
    query.feature = meta.feature[top.dio3rod.peak, ],  n = 50000)

dio3rod.enriched.motifs <- FindMotifs(pr, features = top.dio3rod.peak, 
                              background = peaks.matched)
dio3rod.enriched.motifs <- subset(dio3scone.enriched.motifs, p.adjust < 0.005)

DefaultAssay(pr) <- "chromvar"
dio3rod.diff_act <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "Rods", 
                        only.pos = T, mean.fxn = rowMeans,fc.name = "avg_diff")
dio3rod.diff_act$motif_name <- plyr::mapvalues(x =rownames(dio3rod.diff_act), 
                                       from = motif_list$id, to = motif_list$name)

write.csv(dio3_rod_peak, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3_rod_peak.csv")
write.csv(dio3rod.enriched.motifs, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3rod.enriched.motifs.csv")
write.csv(dio3rod.diff_act, file = "H:/Bob_lab/Christina_dio3/Presentations/dio3rod.diff_act.csv")
rm(open.peaks,meta.feature,peaks.matched,top.dio3rod.peak,annot)
```

Get enriched genes
```{r}
DefaultAssay(pr) <- "RNA"
Idents(pr) <- "Celltype"
scone_genes <- FindMarkers(pr, ident.1 = "S_cone", only.pos = T)
mlcone_genes <- FindMarkers(pr, ident.1 = "ML_cone", only.pos = T)
rod_genes <- FindMarkers(pr, ident.1 = "Rods", only.pos = T)

write.csv(scone_genes, file = "H:/Bob_lab/Christina_dio3/Presentations/scone_genes.csv")
write.csv(mlcone_genes, file = "H:/Bob_lab/Christina_dio3/Presentations/mlcone_genes.csv")
write.csv(rod_genes, file = "H:/Bob_lab/Christina_dio3/Presentations/rod_genes.csv")

dio3_scone_geneDE <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "S_cone", 
                         features = rownames(scone_genes))
dio3_mlcone_geneDE <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "ML_cone", 
                         features = rownames(mlcone_genes))
dio3_rod_geneDE <- FindMarkers(pr,ident.1 = "DIO3_hypo", ident.2 = "Control", 
                         group.by = "genotype", subset.ident = "Rods", 
                         features = rownames(rod_genes))

dio3_rod_peak_geneDE <- intersect(unique(dio3_rod_peak$closest_gene),rownames(dio3_rod_geneDE))
dio3_scone_peak_geneDE <- intersect(unique(dio3_scone_peak$closest_gene),rownames(dio3_scone_geneDE))
dio3_mlcone_peak_geneDE <- intersect(unique(dio3_MLcone_peak$closest_gene),rownames(dio3_mlcone_geneDE))

con_rod_peak_geneDE <- intersect(unique(con_rod_peak$closest_gene),rownames(dio3_rod_geneDE))
con_scone_peak_geneDE <- intersect(unique(con_scone_peak$closest_gene),rownames(dio3_scone_geneDE))
con_mlcone_peak_geneDE <- intersect(unique(con_MLcone_peak$closest_gene),rownames(dio3_mlcone_geneDE))


# Find expression of TFs that have differential cone motif activity
# and differential gene expression
common_con_cone_act <- Reduce(intersect, list(conMLcone.diff_act$motif_name,conscone.diff_act$motif_name))
common_con_cone_act <- strsplit(common_con_cone_act, split="::")
common_con_cone_act <- unlist(common_con_cone_act)
common_con_cone_act <- gsub("\\([^)]+\\)", "", common_con_cone_act)
common_con_cone_act  <- toupper(common_con_cone_act)

common_dio_cone_act <- Reduce(intersect, list(dio3MLcone.diff_act$motif_name,dio3scone.diff_act$motif_name))
common_dio_cone_act <- strsplit(common_dio_cone_act, split="::")
common_dio_cone_act <- unlist(common_dio_cone_act)
common_dio_cone_act <- gsub("\\([^)]+\\)", "", common_dio_cone_act)
common_dio_cone_act  <- toupper(common_dio_cone_act)

intersect (common_dio_cone_act,rownames(dio3_scone_geneDE[dio3_scone_geneDE$avg_log2FC < 0,]) )

```
Link peaks to genes in subset PR object
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
pr <- LinkPeaks(pr, peak.assay = "Peaks", expression.assay = "RNA")
```

Plot ATAC coverage plots
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
gene <- "DIO3"
DefaultAssay(dio3) <- "Peaks"
CoveragePlot(dio3,region = gene, features = gene, extend.upstream = 2000, extend.downstream = 2000,
                         annotation = T,  peaks = T, links = T, 
                         idents = c("Control_S_cone","DIO3_hypo_S_cone","Control_ML_cone","DIO3_hypo_ML_cone"), 
                         group.by = "genotype_celltype2")
```
Perform footprinting analysis on top TF changes from Control in DIO3 hypomorph
```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
pr <- Footprint(pr,
  motif.name = c("MA1109.1"),
  genome = BSgenome.Hsapiens.UCSC.hg38, assay = "Peaks")
```

Save files
```{r}
saveRDS(dio3, file = "F:/Bob_lab/Christina_dio3/R_files/dio3_filtered_macs2_link_int.rds")
```

```{r message=FALSE, warning=FALSE, include=FALSE, results='hide'}
# To update fragment paths
frags <- Fragments(dio3) 
new.paths <- list("F:/Bob_lab/Christina_dio3/Raw/Control/atac_fragments.tsv.gz",
                  "F:/Bob_lab/Christina_dio3/Raw/Dio3/atac_fragments.tsv.gz")
for (i in seq_along(frags)) {
  frags[[i]] <- UpdatePath(frags[[i]], new.path = new.paths[[i]]) 
}
Fragments(dio3) <- NULL
Fragments(dio3) <- frags
rm (frags, new.paths, i)
```




```{r}
sessionInfo()
```
